### PR TITLE
Ember Can

### PR DESCRIPTION
**Pull request by alexbeeken**
Add ability checks to control showing/hiding template links for user roles
Add ability check for route access
Add tests for ember-can template and route checks
Change project model property title to name
Add mirage project, user and project_role endpoints for testing
Add mirage user and project_role factories
Add login check to application route
Add error handling for index and project route
Add unauthorized route and placeholder template
Remove acl.js
Connects to issue #63

### CODE REVIEW

File: app/routes/application.js [hunk 1]
```diff
@@ -12,14 +12,20 @@ export default Ember.Route.extend({
...
+      user: this.store.findRecord('user', this.get('session.currentUserId'))
+    }).catch(() => {
+      return "You are not logged in.";
     });
   },
 
   afterModel(model) {
-    this.set('session.currentUser', model.user);
-    this.set('session.currentUserProjects', model.projects);
+    if (model === "You are not logged in.") {
```
Comment (kscherling) [inline app/routes/application.js]: This seems like a slightly janky way to set this.  Might be better to put a check in the 'beforeModel' hook that redirects before it tries to do stuff....

File: app/routes/index.js [hunk 2]
```diff
@@ -3,14 +3,12 @@ import Ember from 'ember';
 export default Ember.Route.extend({
   session: Ember.inject.service(),
 
-  model() {
-    return Ember.RSVP.hash({
```
Comment (kscherling) [inline app/routes/index.js]: Don't really need to use the RSVP.hash if there is only one property returned, know what I mean?  Simple  `return this.store.find...` should do...

File: app/routes/application.js (hunk 1 above)
Comment (kscherling) [inline app/routes/application.js]: ...or even just transitionTo directly in the catch rather than returning a string than comparing the value of the model in the next hook called...

**Comment by uconnjosh** [comment]
lots of stuff going on here but it looks good to me. maybe you could squash some of your commits though?
```git rebase -i origin/master```

File: app/routes/index.js (hunk 2 above)
Comment (alexbeeken) [inline app/routes/index.js]: I didn't end up using the RSVP it's all done in the beforeModel()

Comment (kscherling) [inline app/routes/index.js]: I'm dumb ;).... red = deleted

File: app/routes/application.js (hunk 1 above)
Comment (kscherling) [inline app/routes/application.js]: I think probably the best route is to use the session.hasCredentials in the before model and bailout before attemping an `store.find` with potentially invalid id/tokens?

File: app/components/nav-user-dropdown/component.js [hunk 3]
```diff
@@ -2,6 +2,11 @@ import layout from './template';
 import RuiDropdownComponent from 'ember-cli-revelation-ui/components/rui-dropdown/component';
 
 export default RuiDropdownComponent.extend({
+  session: Ember.inject.service(),
+  currentProject: Ember.inject.service(),
+  user: Ember.computed.reads('session.currentUser'),
+  role: Ember.computed.reads('currentProject.currentRole'),
```
Comment (smashedlife) [inline app/components/nav-user-dropdown/component.js]: Weren't we alphabetizing these?

File: app/mirage/config.js [hunk 4]
```diff
@@ -38,6 +38,60 @@ export default function() {
...
+    let user = db.users.find(id);
+
+    return {
+      data: {
+        type: 'users',
+        id: id,
+        attributes: {
+         avatar: user.avatar,
+         email: user.email,
+         "is-superuser": user.isSuperuser,
```
Comment (smashedlife) [inline app/mirage/config.js]: I thought we were removing is superuser from these params

File: app/models/user.js [hunk 5]
```diff
@@ -1,9 +1,9 @@
 import DS from 'ember-data';
 
 export default DS.Model.extend({
+  lastVisitedProjectId: DS.attr('Number'),
   avatar:               DS.attr('String'),
   email:                DS.attr('String'),
   isSuperuser:          DS.attr('Boolean'),
-  lastVisitedProjectId: DS.attr('String'),
```
Comment (smashedlife) [inline app/models/user.js]: Shouldn't this be alphabetized?

File: app/routes/index.js [hunk 6]
```diff
@@ -3,14 +3,12 @@ import Ember from 'ember';
...
-      user: this.store.findRecord('user', this.get('session.currentUser'))
-    });
-  },
-
-  beforeModel(transition) {
-    this.transitionTo('project', this.get('session.currentUser.lastVisitedProjectId'));
+  beforeModel() {
+    if (this.get('session.currentUser.lastVisitedProjectId')) {
```
Comment (smashedlife) [inline app/routes/index.js]: I am not sure if I understand this correctly. Shouldn't we be checking for access rights and not just if they have lastVisitedProjectID as a value to determine if they should be authorized or not

File: app/routes/project/overview/index.js [hunk 7]
```diff
@@ -1,7 +1,12 @@
 import Ember from 'ember';
+import { CanMixin } from 'ember-can';
 
-export default Ember.Route.extend({
+export default Ember.Route.extend(CanMixin, {
   beforeModel() {
-    this.transitionTo('project.overview.activities');
+    if (!this.can('see overview in navigation')) {
```
Comment (smashedlife) [inline app/routes/project/overview/index.js]: Instead of swapping a false statement around, wouldn't it be easier to start with the truth and then go to unauthorized?

File: app/components/nav-user-dropdown/component.js (hunk 3 above)
Comment (alexbeeken) [inline app/components/nav-user-dropdown/component.js]: yep I'll push it up in a sec

File: app/routes/project/setup/index.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import Ember from 'ember';
+import { CanMixin } from 'ember-can';
+
+export default Ember.Route.extend(CanMixin, {
+  // Temporarily added since rails toures don't exist yet
+  beforeModel() {
+    if (!this.can('see project setup in navigation')) {
```
Comment (smashedlife) [inline app/routes/project/setup/index.js]: Same here around the truth first and then the unauthorized statement. I can understand if we had a guard clause.

File: app/routes/project/setup/settings/digests/index.js [hunk 9]
```diff
@@ -1,7 +1,13 @@
 import Ember from 'ember';
+import { CanMixin } from 'ember-can';
+
+export default Ember.Route.extend(CanMixin, {
 
-export default Ember.Route.extend({
   beforeModel() {
-    this.transitionTo('project.setup.settings.digests.options');
+    if (!this.can('see project setup in navigation')) {
```
Comment (smashedlife) [inline app/routes/project/setup/settings/digests/index.js]: Same here around the truth first and then the unauthorized statement.

File: app/routes/project/setup/settings/digests/options.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import Ember from 'ember';
+import { CanMixin } from 'ember-can';
+
+export default Ember.Route.extend(CanMixin, {
+
+  beforeModel() {
+    if (!this.can('see project setup in navigation')) {
```
Comment (smashedlife) [inline app/routes/project/setup/settings/digests/options.js]: Same here around the truth first and then the unauthorized statement. I can understand if we had a guard clause.

File: app/routes/project/setup/settings/digests/preview.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import Ember from 'ember';
+import { CanMixin } from 'ember-can';
+
+export default Ember.Route.extend(CanMixin, {
+
+  beforeModel() {
+    if (!this.can('see project setup in navigation')) {
```
Comment (smashedlife) [inline app/routes/project/setup/settings/digests/preview.js]: Same here around the truth first and then the unauthorized statement.

File: app/routes/project/setup/settings/index.js [hunk 12]
```diff
@@ -1,8 +1,13 @@
 import Ember from 'ember';
+import { CanMixin } from 'ember-can';
 
-export default Ember.Route.extend({
+export default Ember.Route.extend(CanMixin, {
   // Temporarily added since rails toures don't exist yet
   beforeModel() {
-    this.transitionTo('project.setup.settings.digests');
+    if (!this.can('see project setup in navigation')) {
```
Comment (smashedlife) [inline app/routes/project/setup/settings/index.js]: Same here around the truth first and then the unauthorized statement.

File: app/services/current-project.js [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import Ember from 'ember';
+
+export default Ember.Service.extend({
+  session: Ember.inject.service(),
+  store: Ember.inject.service(),
+  projectId: null,
+  currentRole: null
```
Comment (smashedlife) [inline app/services/current-project.js]: Shouldn't this be alphabetized?

File: app/templates/unauthorized.hbs [hunk 14]
```diff
@@ -0,0 +1,11 @@
+<div class='center-fix'>
```
Comment (smashedlife) [inline app/templates/unauthorized.hbs]: Lets use the default one for Rails and also bounce a 401 error code.

**Comment by smashedlife** [comment]
Just curious, why did we load the SVG logos into test public?

File: app/routes/index.js (hunk 6 above)
Comment (alexbeeken) [inline app/routes/index.js]: Do you think it makes sense to have a canAccessEntirePage ability? I felt like the auth_token was pretty much the access rights check at this level.

Comment (alexbeeken) [inline app/routes/index.js]: Like once you are authorized you will not need to check access rights until you are inside a project.

Comment (smashedlife) [inline app/routes/index.js]: How would you imagine access being given to a user that say didn't have access to see the email vs someone who did through the auth token?

File: app/templates/unauthorized.hbs (hunk 14 above)
Comment (alexbeeken) [inline app/templates/unauthorized.hbs]: Can I do that in another pull request?

Comment (smashedlife) [inline app/templates/unauthorized.hbs]: Is there a reason to do it in another pull request vs the one you have currently open?

Comment (alexbeeken) [inline app/templates/unauthorized.hbs]: The stylings for the rails page uses whitelabel. It's going to take time to make it all work in ember-admin.

Comment (smashedlife) [inline app/templates/unauthorized.hbs]: The unauthorized page isn't whitelabel. View the 401 page in the public folder.
